aioxmpp/avatar/service.py

```diff
@@ -748,33 +807,107 @@ def publish_avatar_set(self, avatar_set):
-                                            avatar_set.image_bytes)
-                    self._vcard_id = avatar_set.png_id
-                    yield from self._vcard.set_vcard(my_vcard)
-                    yield from self._presence_server.resend_presence()
+                done = True
+
+            if self._synchronize_vcard:
+                my_vcard = yield from self._vcard.get_vcard()
+                my_vcard.set_photo_data("image/png",
+                                        avatar_set.image_bytes)
+                self._vcard_id = avatar_set.png_id
+                yield from self._vcard.set_vcard(my_vcard)
+                yield from self._presence_server.resend_presence()
+                done = True
+
+        if not done:
+            raise RuntimeError(
+                "failed to publish avatar: no protocol available"
+            )
+
+    @asyncio.coroutine
+    def _disable_vcard_avatar(self):
+        my_vcard = yield from self._vcard.get_vcard()
+        my_vcard.clear_photo_data()
+        self._vcard_id = ""
+        yield from self._vcard.set_vcard(my_vcard)
+        yield from self._presence_server.resend_presence()
+
 
     @asyncio.coroutine
     def disable_avatar(self):
         """
         Temporarily disable the avatar.
 
-        This is done by setting the avatar metadata node empty.
+        If :attr:`synchronize_vcard` is true, the vCard avatar is
+        disabled even if disabling the PEP avatar fails.
+
+        This is done by setting the avatar metadata node empty and if
+        :attr:`synchronize_vcard` is true, downloading the vCard,
+        removing the avatar data and reuploading the vCard.
+
+        :raises RuntimeError: if an exception is raised by the spawned
+            tasks.
         """
 
         with (yield from self._publish_lock):
-            try:
-                yield from self._pep.publish(
+            todo = []
+            if self._synchronize_vcard:
+                todo.append(self._disable_vcard_avatar())
+
+            if (yield from self._pep.available()):
+                todo.append(self._pep.publish(
                     namespaces.xep0084_metadata,
                     avatar_xso.Metadata()
-                )
-            finally:
-                if self._synchronize_vcard:
-                    my_vcard = yield from self._vcard.get_vcard()
-                    my_vcard.clear_photo_data()
-                    self._vcard_id = ""
-                    yield from self._vcard.set_vcard(my_vcard)
-                    yield from self._presence_server.resend_presence()
+                ))
+
+            done, _ = yield from asyncio.wait(todo)
+            exceptions = []
+            for fut in done:
+                if fut.exception() is not None:
+                    exceptions.append(fut.exception())
+                fut.result() # to avoid warnings
+            if exceptions:
+                raise RuntimeError from exceptions[0]
```

Comment:
I really wish python had a built-in way to re-raise multiple exceptions :(.

In any case, this exception deserves a text. I’d suggest ``"Failed to disable avatar ({})".format(", ".join(map(repr, exceptions)))``